ethcontract/src/batch.rs

```diff
@@ -49,40 +49,44 @@ impl<T: Web3BatchTransport> CallBatch<T> {
         }
     }
 
-    /// Execute and resolve all enqueued CallRequests in a single RPC call
-    pub async fn execute_all(self) -> Result<(), Web3Error> {
-        if self.requests.is_empty() {
-            return Ok(());
+    /// Execute and resolve all enqueued CallRequests in a batched RPC call, `chunk_size` requests per roundtrip.
+    /// Top level request failures will be forwarded to the individual requests.
+    pub async fn execute_all(self, batch_size: usize) {
+        let Self { inner, requests } = self;
+        let mut iterator = requests.into_iter().peekable();
+        while iterator.peek().is_some() {
+            let (requests, senders): (Vec<_>, Vec<_>) = iterator.by_ref().take(batch_size).unzip();
```

Comment:
Maybe you can use the `chunks` method.

Reply:
Nice suggestion!

Reply:
I fought with this for quite a bit yesterday. The problem with chunks is that it gives you slices on the underlying vector. From that I wasn't able to consume the elements (they are hiding behind the slice reference) which is needed because the `sender` of the result propagation channel is a "one shot" (not clone-able) which consumes itself upon send.

So I really need to consume the items of `self.requests`. Turning it into an iterator and taking from it was the cleanest way I found to make it work. Happy to see some pseudo code for how to make it work with chunks. I might have missed something.

Reply:
That makes sense and your solution is fine.